File: src/leetcode每日一题/_03_23/440. 字典序的第K小数字.java
package leetcode每日一题._03_23;

class Solution {
    public int findKthNumber(int n, int k) {
        int cur = 1;
        k--;
        while(k > 0) {
            long cnts = dfs(cur, cur, n);
            if(cnts <= k) {
                k -= cnts;
                cur++;
            } else {
                k--;
                cur *= 10;
            }
        }
        return cur;
    }

    private long dfs(long l, long r, int n) {
        if(l > n)
            return 0;
        return Math.min(r, n) - l + 1 + dfs(l * 10, r * 10 + 9, n);
    }
}

File: src/leetcode每日一题/_03_23/440. 字典序的第K小数字.go
package _03_23

func findKthNumber(n int, k int) int {
	var dfs func(l, r int) int
	dfs = func(l, r int) int {
		if l > n {
			return 0
		}
		if r > n {
			r = n
		}
		return r - l + 1 + dfs(l*10, r*10+9)
	}

	cur := 1
	k--
	for k > 0 {
		cnts := dfs(cur, cur)
		if cnts <= k {
			k -= cnts
			cur++
		} else {
			k--
			cur *= 10
		}
	}
	return cur
}
